File: routes/voice.js
const express = require('express');
const twilio = require('twilio');
const router = express.Router();
const llmAgent = require('./../services/llmAgent');
const patientService = require('./../services/patientService');

// In-memory conversation state keyed by CallSid. Fine for a single-instance demo deployment;
// see README "Known Limitations" for how this would need to change for multi-instance hosting.
const callStates = new Map();

const VOICE = process.env.TWILIO_TTS_VOICE || 'Polly.Joanna';
const GATHER_TIMEOUT_MESSAGE = "I didn't catch that. Are you still there?";

function buildGatherResponse({ say, hangup }) {
  const twiml = new twilio.twiml.VoiceResponse();

  if (hangup) {
    twiml.say({ voice: VOICE }, say);
    twiml.hangup();
    return twiml;
  }

  const gather = twiml.gather({
    input: 'speech',
    action: '/voice/gather',
    method: 'POST',
    speechTimeout: 'auto',
    speechModel: 'phone_call'
  });
  gather.say({ voice: VOICE }, say);

  // If Gather gets no speech at all, Twilio falls through here.
  twiml.say({ voice: VOICE }, GATHER_TIMEOUT_MESSAGE);
  twiml.redirect({ method: 'POST' }, '/voice/incoming?retry=1');

  return twiml;
}

// Entry point for inbound calls
router.post('/incoming', async (req, res) => {
  const callSid = req.body.CallSid;
  console.log(`[voice] Incoming call ${callSid} from ${req.body.From}`);

  let state = callStates.get(callSid);
  if (!state) {
    state = llmAgent.createCallState();
    callStates.set(callSid, state);
  }

  try {
    const isRetry = req.query.retry === '1';
    const kickoff = isRetry
      ? 'The caller did not respond. Gently check if they are still on the line and continue helping them register.'
      : 'The call just started. Greet the caller warmly, introduce yourself as the intake assistant, and ask how you can help (registering as a new patient).';

    const { say, hangup } = await llmAgent.runTurn(state, kickoff);
    const twiml = buildGatherResponse({ say, hangup });
    res.type('text/xml').send(twiml.toString());
  } catch (err) {
    console.error('[voice/incoming] error:', err);
    const twiml = new twilio.twiml.VoiceResponse();
    twiml.say({ voice: VOICE }, "I'm sorry, we're experiencing a technical issue. Please call back in a few minutes.");
    twiml.hangup();
    res.type('text/xml').send(twiml.toString());
  }
});

// Handles each turn of speech input
router.post('/gather', async (req, res) => {
  const callSid = req.body.CallSid;
  const speechResult = req.body.SpeechResult || '';
  console.log(`[voice] ${callSid} said: "${speechResult}"`);

  let state = callStates.get(callSid);
  if (!state) {
    state = llmAgent.createCallState();
    callStates.set(callSid, state);
  }

  try {
    const { say, hangup } = await llmAgent.runTurn(state, speechResult || '(no speech detected, please re-prompt)');
    const twiml = buildGatherResponse({ say, hangup });
    res.type('text/xml').send(twiml.toString());

    if (hangup) {
      finalizeCall(callSid, state);
    }
  } catch (err) {
    console.error('[voice/gather] error:', err);
    const twiml = new twilio.twiml.VoiceResponse();
    twiml.say({ voice: VOICE }, "I'm sorry, something went wrong on our end saving your information. Please try calling back shortly.");
    twiml.hangup();
    res.type('text/xml').send(twiml.toString());
    finalizeCall(callSid, state);
  }
});

// Call status callback (completed, failed, busy, no-answer) - cleans up state on drop/hangup
router.post('/status', (req, res) => {
  const callSid = req.body.CallSid;
  const callStatus = req.body.CallStatus;
  console.log(`[voice/status] ${callSid} -> ${callStatus}`);

  if (['completed', 'failed', 'busy', 'no-answer', 'canceled'].includes(callStatus)) {
    const state = callStates.get(callSid);
    if (state) {
      finalizeCall(callSid, state);
    }
  }
  res.status(200).send('OK');
});

function finalizeCall(callSid, state) {
  try {
    if (state.transcript.length > 0) {
      const transcriptText = state.transcript.map((t) => `${t.role}: ${t.text}`).join('\n');
      patientService.saveTranscript({
        callSid,
        patientId: state.savedPatientId,
        transcript: transcriptText
      });
      console.log(`[voice] Call ${callSid} final payload:`, JSON.stringify({
        patientId: state.savedPatientId,
        matchedExistingPatientId: state.matchedExistingPatientId
      }));
    }
  } catch (err) {
    console.error('[voice] Failed to persist transcript:', err);
  } finally {
    callStates.delete(callSid);
  }
}

module.exports = router;

File: app/routes/voice.py
import os

from flask import Blueprint, Response, request
from twilio.twiml.voice_response import VoiceResponse

from app import patient_service, voice_agent

bp = Blueprint("voice", __name__, url_prefix="/voice")

# In-memory conversation state keyed by CallSid. Fine for a single-instance demo deployment;
# see README "Known Limitations" for how this would need to change for multi-instance hosting.
call_states = {}

VOICE = os.environ.get("TWILIO_TTS_VOICE", "Polly.Joanna")
GATHER_TIMEOUT_MESSAGE = "I didn't catch that. Are you still there?"


def _build_gather_response(say, hangup):
    twiml = VoiceResponse()

    if hangup:
        twiml.say(say, voice=VOICE)
        twiml.hangup()
        return twiml

    gather = twiml.gather(
        input="speech",
        action="/voice/gather",
        method="POST",
        speech_timeout="auto",
        speech_model="phone_call",
    )
    gather.say(say, voice=VOICE)

    twiml.say(GATHER_TIMEOUT_MESSAGE, voice=VOICE)
    twiml.redirect("/voice/incoming?retry=1", method="POST")

    return twiml


def _finalize_call(call_sid, state):
    try:
        if state["transcript"]:
            transcript_text = "\n".join(f"{t['role']}: {t['text']}" for t in state["transcript"])
            patient_service.save_transcript(call_sid, state.get("saved_patient_id"), transcript_text)
            print(
                f"[voice] Call {call_sid} final payload:",
                {
                    "patientId": state.get("saved_patient_id"),
                    "matchedExistingPatientId": state.get("matched_existing_patient_id"),
                },
            )
    except Exception as err:  # noqa: BLE001
        print("[voice] Failed to persist transcript:", err)
    finally:
        call_states.pop(call_sid, None)


@bp.post("/incoming")
def incoming():
    call_sid = request.form.get("CallSid")
    print(f"[voice] Incoming call {call_sid} from {request.form.get('From')}")

    state = call_states.setdefault(call_sid, voice_agent.create_call_state())

    try:
        is_retry = request.args.get("retry") == "1"
        kickoff = (
            "The caller did not respond. Gently check if they are still on the line and continue helping them register."
            if is_retry
            else "The call just started. Greet the caller warmly, introduce yourself as the intake assistant, and ask how you can help (registering as a new patient)."
        )
        result = voice_agent.run_turn(state, kickoff)
        twiml = _build_gather_response(result["say"], result["hangup"])
        return Response(str(twiml), mimetype="text/xml")
    except Exception as err:  # noqa: BLE001
        print("[voice/incoming] error:", err)
        twiml = VoiceResponse()
        twiml.say("I'm sorry, we're experiencing a technical issue. Please call back in a few minutes.", voice=VOICE)
        twiml.hangup()
        return Response(str(twiml), mimetype="text/xml")


@bp.post("/gather")
def gather():
    call_sid = request.form.get("CallSid")
    speech_result = request.form.get("SpeechResult", "")
    print(f'[voice] {call_sid} said: "{speech_result}"')

    state = call_states.setdefault(call_sid, voice_agent.create_call_state())

    try:
        result = voice_agent.run_turn(state, speech_result or "(no speech detected, please re-prompt)")
        twiml = _build_gather_response(result["say"], result["hangup"])
        response = Response(str(twiml), mimetype="text/xml")
        if result["hangup"]:
            _finalize_call(call_sid, state)
        return response
    except Exception as err:  # noqa: BLE001
        print("[voice/gather] error:", err)
        twiml = VoiceResponse()
        twiml.say(
            "I'm sorry, something went wrong on our end saving your information. Please try calling back shortly.",
            voice=VOICE,
        )
        twiml.hangup()
        _finalize_call(call_sid, state)
        return Response(str(twiml), mimetype="text/xml")


@bp.post("/status")
def status():
    call_sid = request.form.get("CallSid")
    call_status = request.form.get("CallStatus")
    print(f"[voice/status] {call_sid} -> {call_status}")

    if call_status in ("completed", "failed", "busy", "no-answer", "canceled"):
        state = call_states.get(call_sid)
        if state:
            _finalize_call(call_sid, state)

    return "OK", 200
